app/rdf/query-dimension-values.ts

```diff
@@ -1,29 +1,133 @@
+  value: string | number,
+  dimension: CubeDimension
+) => {
+  if (!dimension.datatype) {
+    return `<${value}>`;
+  } else {
+    // Seems fragile
+    if (dimension.datatype.value === ns.xsd.gYear.value) {
+      return `"${value}"^^xsd:gYear`;
+    } else if (dimension.datatype.value === ns.xsd.date.value) {
+      return `"${value}"^^xsd:date`;
+    } else if (dimension.datatype.value === ns.xsd.dateTime.value) {
+      return `"${value}"^^xsd:dateTime`;
+    } else {
+      return `"${value}"`;
+    }
+  }
+};
+
+const formatFilterIntoSparqlFilter = (
+  filter: Filters[string],
+  dimension: CubeDimension,
+  versioned: boolean,
+  index: number
+) => {
+  const suffix = versioned ? "_unversioned" : "";
+  if (filter.type === "single") {
+    return `FILTER ( (?dimension${suffix}${index} = ${formatFilterValue(
+      filter.value,
+      dimension
+    )}) )`;
+  } else if (filter.type === "multi") {
+    return `FILTER ( (?dimension${suffix}${index} in (${Object.keys(
+      filter.values
+    )
+      .map((x) => formatFilterValue(x, dimension))
+      .join(",")}) ) )`;
+  } else {
+    return "";
+  }
+};
+
 /**
  * Load dimension values.
+ *
+ * Filters on other dimensions can be passed.
+ *
  */
-export async function loadDimensionValues({
-  datasetIri,
-  dimensionIri,
-}: {
-  datasetIri: Term | undefined;
-  dimensionIri: Term | undefined;
-}): Promise<Array<Literal | NamedNode>> {
-  const query = SELECT.DISTINCT`?value`.WHERE`
-    ${datasetIri} ${cube.observationSet} ?observationSet .
-    ?observationSet ${cube.observation} ?observation .
+export async function loadDimensionValues(
+  {
+    datasetIri,
+    dimension,
+    cube,
+  }: {
+    datasetIri: Term | undefined;
+    dimension: CubeDimension;
+    cube: Cube;
+  },
+  filters?: Filters
+): Promise<Array<Literal | NamedNode>> {
+  const dimensionIri = dimension.path;
+
+  let filterList = filters ? Object.entries(filters) : [];
+  filterList = filterList.slice(
+    0,
+    filterList.findIndex(([iri]) => iri == dimensionIri?.value) + 1
+  );
+
+  let query = SELECT.DISTINCT`?value`.WHERE`
+    ${datasetIri} ${cubeNs.observationSet} ?observationSet .
+    ?observationSet ${cubeNs.observation} ?observation .
     ?observation ${dimensionIri} ?value .
+    ${
+      filters
+        ? filterList
+            .map(([iri, value], idx) => {
+              const filterDimension = cube.dimensions.find(
+                (d) => d.path?.value === iri
+              );
+              if (
+                !filterDimension ||
+                value.type === "range" ||
+                dimensionIri?.value === iri
+              ) {
+                return "";
+              }
+              const versioned = filterDimension
+                ? dimensionIsVersioned(filterDimension)
+                : false;
+              return `${
+                versioned
+                  ? `?dimension${idx} <http://schema.org/sameAs> ?dimension_unversioned${idx}.`
```

Comment:
I think you also had an issue with using ${ns.schema.sameAs} here? I faced it in geoCoordinates query, you can take a look: https://github.com/visualize-admin/visualization-tool/blob/main/app/rdf/query-geo-coordinates.ts#L31. Not sure if this is an elegant solution, but it does a trick :)

Reply:
Mmh yes but my problem is that if I use the ns.schema.sameAs.toString(), it will correctly expand it to `<https://...>` but my problem is that the prefix will not be added to the query since the usage of the sparql bit will not be detected (since it is nested and already resolved).

```
const { SELECT, sparql } = require("@tpluscode/sparql-builder");
const { schema } = require("@tpluscode/rdf-ns-builders");

const filters = ["a"];
const query = SELECT.DISTINCT`?value`.WHERE`
    ${filters.map(
      (f) =>
        `FILTER (( ${sparql`${schema.sameAs}`.toString({
          prologue: true,
        })}) = 1 )`
    )}
  `;

console.log(query.build());
```

```
SELECT DISTINCT ?value

WHERE {

    FILTER (( schema:sameAs) = 1 )
  
}
```

Here the prefix is not added.

I will post an issue on sparql builder

Reply:
Hmmm, right, I see that I successfully included` ${ns.schema.[smth]}` in geo coords query, so it was already appended in that case.  Good 👍

Reply:
https://github.com/tpluscode/sparql-builder/issues/62 ;)

Reply:
Nice :D